File: actool/packer.py
"""
Atlas packing for CAR files.

Implements a column-first bin-packing algorithm that matches Apple's actool
output. Images are grouped by pixel format and scale, then packed into
atlas textures.

Packing rules:
- 2px margin on all edges, 2px gap between images
- Sorted by height descending, then width descending
- Column-based: stack images vertically in columns
- When a column is full, start a new column to the right
- Minimum 2 images per format/scale group to trigger packing
"""

from dataclasses import dataclass, field

from PIL import Image


MARGIN = 2
GAP = 2


@dataclass
class PackedImage:
    """An image placed in an atlas."""
    name: str
    identifier: int
    width: int
    height: int
    x: int = 0
    y: int = 0
    pixel_data: bytes = b""
    pixel_format: bytes = b"BGRA"
    scale: int = 1
    is_template: bool = False
    template_rendering_intent: int = 4  # bitmapEncoding: 0=original, 4=automatic, 2=template
    part: int = 181  # PART_REGULAR by default, PART_ICON for icons
    dim2: int = 0  # For icon images, the multisize index
    appearance: int = 0
    direction: int = 0


@dataclass
class Atlas:
    """A packed atlas texture containing multiple images."""
    width: int = 0
    height: int = 0
    pixel_format: bytes = b"BGRA"
    scale: int = 1
    dim1: int = 0
    images: list[PackedImage] = field(default_factory=list)
    pixel_data: bytes = b""

    @property
    def name(self) -> str:
        fmt_idx = 0 if self.pixel_format == b"BGRA" else 1
        return f"ZZZZPackedAsset-{self.scale}.{self.dim1}.{fmt_idx}-gamut0"

    @property
    def bytes_per_row(self) -> int:
        """Row stride in bytes for the encoded buffer, aligned to 32 bytes.

        Uses the actual bytes-per-pixel of the atlas format (4 for BGRA,
        2 for GA8). vImage's deepmap2 encoder reads `width * pixel_size`
        bytes per row at this stride.
        """
        bpp = 4 if self.pixel_format == b"BGRA" else 2
        exact = self.width * bpp
        return ((exact + 31) // 32) * 32

    def render(self):
        """Render all packed images into a single atlas pixel buffer.

        Rows are padded to 32-byte alignment to match the encoder's
        expected stride. The buffer uses top-down row order; the INLK
        y coordinates are flipped to bottom-left origin separately by
        the compiler.

        Sub-images are placed at byte offsets matching the atlas's
        actual bytes-per-pixel (2 for GA8, 4 for BGRA), not at a fixed
        4-bpp positioning. This keeps the GA8 atlas consistent so that
        vImage encodes the full pixel content.
        """
        bpp = 4 if self.pixel_format == b"BGRA" else 2
        bpr = self.bytes_per_row
        buf = bytearray(bpr * self.height)

        for img in self.images:
            src_stride = img.width * bpp

            for row in range(img.height):
                src_off = row * src_stride
                dst_off = (img.y + row) * bpr + img.x * bpp
                buf[dst_off:dst_off + src_stride] = \
                    img.pixel_data[src_off:src_off + src_stride]

        self.pixel_data = bytes(buf)


def pack_images(images: list[PackedImage], max_width: int = 2048) -> Atlas:
    """Pack images into atlas(es) using shelf-based packing.

    Uses shelf packing to match Apple's actool layout: images are arranged
    in horizontal shelves, with column stacking within each shelf. When a
    single atlas would exceed the height limit, splits into multiple atlases.

    Returns a single Atlas when all images fit, or the first atlas from
    a multi-atlas split.
    """
    if not images:
        return Atlas()

    atlases = pack_images_split(images, max_width=max_width)
    return atlases[0] if atlases else Atlas()


def pack_images_split(images: list[PackedImage], max_width: int = 2048,
                      max_height: int = 196) -> list[Atlas]:
    """Pack images into multiple atlases using shelf-based bin packing.

    Matches Apple's actool packing strategy:
    - First shelf determines atlas width
    - Subsequent shelves fill within that width
    - Images can be stacked vertically within a shelf (column-within-shelf)
    - When height exceeds max_height, overflow images go to a new atlas
    - When starting a new atlas, widest remaining image goes first
    """
    if not images:
        return []

    remaining = sorted(images, key=lambda i: (-i.height, -i.width))
    atlases = []

    while remaining:
        atlas = Atlas(
            pixel_format=remaining[0].pixel_format,
            scale=remaining[0].scale,
        )

        placed, overflow = _pack_shelf_atlas(
            atlas, remaining, max_width, max_height)
        atlases.append(atlas)
        remaining = overflow

    return atlases


def _pack_shelf_atlas(atlas: Atlas, sorted_imgs: list[PackedImage],
                      max_width: int, max_height: int
                      ) -> tuple[list[PackedImage], list[PackedImage]]:
    """Pack images into a single atlas using shelf-based packing.

    The first shelf can grow the atlas width freely. Subsequent shelves
    are constrained to fit within the width established by the first shelf.

    Returns (placed, overflow) lists.
    """
    # Shelves: each shelf has columns of images
    shelves: list[dict] = []  # {y, height, columns: [{x, width, images, bottom}]}
    atlas_width = 0  # Set by first shelf, constrains subsequent shelves

    placed = []
    overflow = []

    for img in sorted_imgs:
        fit = False

        # Try to fit in an existing shelf's column or as a new column
        for si, shelf in enumerate(shelves):
            is_first_shelf = (si == 0)

            # Try existing columns within this shelf
            for col in shelf['columns']:
                if img.width <= col['width']:
                    new_bottom = col['bottom'] + GAP + img.height
                    if new_bottom <= shelf['y'] + shelf['height']:
                        # Fits within shelf height
                        img.x = col['x']
                        img.y = col['bottom'] + GAP
                        col['images'].append(img)
                        col['bottom'] = img.y + img.height
                        placed.append(img)
                        fit = True
                        break

            if fit:
                break

            # Try new column on this shelf
            if shelf['columns']:
                last_col = shelf['columns'][-1]
                new_x = last_col['x'] + last_col['width'] + GAP
            else:
                new_x = MARGIN

            if img.height <= shelf['height']:
                # First shelf can grow atlas width; others are constrained
                width_ok = (is_first_shelf and
                            new_x + img.width + MARGIN <= max_width)
                if not width_ok and atlas_width > 0:
                    width_ok = new_x + img.width + MARGIN <= atlas_width
                if not width_ok and not shelf['columns']:
                    width_ok = True  # First column always fits

                if width_ok:
                    img.x = new_x
                    img.y = shelf['y']
                    shelf['columns'].append({
                        'x': new_x,
                        'width': img.width,
                        'images': [img],
                        'bottom': img.y + img.height,
                    })
                    # First shelf determines atlas width
                    new_right = new_x + img.width + MARGIN
                    if new_right > atlas_width:
                        atlas_width = new_right
                    placed.append(img)
                    fit = True
                    break

        if fit:
            continue

        # Try new shelf
        if shelves:
            last_shelf = shelves[-1]
            new_y = last_shelf['y'] + last_shelf['height'] + GAP
        else:
            new_y = MARGIN

        if new_y + img.height + MARGIN <= max_height or not shelves:
            shelf_height = img.height
            img.x = MARGIN
            img.y = new_y

            shelf = {
                'y': new_y,
                'height': shelf_height,
                'columns': [{
                    'x': MARGIN,
                    'width': img.width,
                    'images': [img],
                    'bottom': img.y + img.height,
                }],
            }
            shelves.append(shelf)

            # Update atlas width
            new_right = MARGIN + img.width + MARGIN
            if new_right > atlas_width:
                atlas_width = new_right
            placed.append(img)
        else:
            overflow.append(img)

    # Calculate atlas dimensions
    if shelves:
        atlas.width = atlas_width
        max_bottom = 0
        for shelf in shelves:
            for col in shelf['columns']:
                if col['bottom'] > max_bottom:
                    max_bottom = col['bottom']
        atlas.height = max_bottom + MARGIN
    else:
        atlas.width = 0
        atlas.height = 0

    atlas.images = placed
    return placed, overflow


def group_for_packing(renditions) -> tuple[list, list]:
    """Group renditions into packable groups and inline renditions.

    Returns (pack_groups, inline_renditions) where pack_groups is a list of
    (format, scale, renditions_list) tuples.
    """
    from . import car

    # Group by (pixel_format, scale)
    groups: dict[tuple[bytes, int], list] = {}
    force_inline = []

    # Threshold: icon images >= 256px are stored inline
    ICON_INLINE_THRESHOLD = 256
    # Atlas packing limits (images exceeding these go inline)
    PACK_MAX_WIDTH = 262
    PACK_MAX_HEIGHT = 196
    PACK_MARGIN = 4  # 2px margin on each side

    for rend in renditions:
        # Multisize, color, and data renditions are always inline
        if rend.layout in (car.LAYOUT_MULTISIZE_IMAGE, car.LAYOUT_COLOR,
                           car.LAYOUT_RAW_DATA, car.LAYOUT_METADATA):
            force_inline.append(rend)
            continue
        # Large app icon renditions are stored inline
        if rend.part == car.PART_ICON and rend.width >= ICON_INLINE_THRESHOLD:
            force_inline.append(rend)
            continue
        # Renditions with CSI overrides (pre-built) are always inline
        if hasattr(rend, '_csi_override'):
            force_inline.append(rend)
            continue
        # Images too large for atlas packing are stored inline
        if (rend.width >= PACK_MAX_WIDTH - PACK_MARGIN or
                rend.height >= PACK_MAX_HEIGHT - PACK_MARGIN):
            force_inline.append(rend)
            continue
        # Group by format, scale, and sprite atlas
        # Icons pack together with regular images (Apple doesn't separate them)
        atlas_id = rend.sprite_atlas_id  # 0 for regular, non-zero for sprites
        key = (rend.pixel_format, rend.scale, atlas_id)
        if key not in groups:
            groups[key] = []
        groups[key].append(rend)

    # Only pack groups with 2+ distinct imagesets (facets).
    # Appearance variants of the same imageset share an identifier,
    # and the system actool requires at least 2 distinct imagesets
    # before packing into an atlas.
    pack_groups = []
    inline = list(force_inline)

    for (fmt, scale, _atlas_id), rends in sorted(groups.items()):
        distinct_facets = len({r.identifier for r in rends})
        if distinct_facets >= 2:
            pack_groups.append((fmt, scale, rends))
        else:
            inline.extend(rends)

    return pack_groups, inline

File: src/packer.rs
//! Atlas packing for CAR files.
//!
//! Shelf-based bin packing matching Apple's actool layout: images are
//! arranged in horizontal shelves, with column stacking within each shelf.
//! When a single atlas would exceed the height limit, remaining images
//! overflow into additional atlases.

pub const MARGIN: u32 = 2;
pub const GAP: u32 = 2;

pub const PART_REGULAR: u32 = 181;
pub const PART_ICON: u32 = 184;

#[derive(Debug, Clone)]
pub struct PackedImage {
    pub name: String,
    pub identifier: u32,
    pub width: u32,
    pub height: u32,
    pub x: u32,
    pub y: u32,
    pub pixel_data: Vec<u8>,
    pub pixel_format: [u8; 4],
    pub scale: u32,
    pub is_template: bool,
    /// bitmapEncoding: 0=original, 4=automatic, 2=template. Matches Python's
    /// default of 4 (automatic).
    pub template_rendering_intent: i32,
    pub part: u32,
    pub dim2: u32,
    pub appearance: u32,
    pub direction: u32,
    /// True when the source was a vector (SVG / PDF) rasterized into
    /// pixels. CoreUI sets a CSI flag bit (0x04) for these so the runtime
    /// knows the rendition originated from a vector mask.
    pub is_svg_rasterization: bool,
    /// Attribute 24 — appearance-variant axis. Variant=1 images route
    /// into a separate atlas (gamut=1) per Apple's layout.
    pub variant: u32,
}

impl PackedImage {
    pub fn new(name: String, identifier: u32, width: u32, height: u32) -> Self {
        Self {
            name,
            identifier,
            width,
            height,
            x: 0,
            y: 0,
            pixel_data: Vec::new(),
            pixel_format: *b"BGRA",
            scale: 1,
            is_template: false,
            template_rendering_intent: 4,
            part: PART_REGULAR,
            dim2: 0,
            appearance: 0,
            direction: 0,
            is_svg_rasterization: false,
            variant: 0,
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct Atlas {
    pub width: u32,
    pub height: u32,
    pub pixel_format: [u8; 4],
    pub scale: u32,
    pub dim1: u32,
    pub images: Vec<PackedImage>,
    pub pixel_data: Vec<u8>,
    /// Apple's appearance-specialization axis baked into the atlas name
    /// and tracked as attribute 24 on the atlas's rendition key. 0 for
    /// the primary atlas; 1 for the alternate variant emitted alongside
    /// when icon.json declares top-level fill-specializations.
    pub gamut: u32,
}

impl Atlas {
    pub fn name(&self) -> String {
        // The third numeric component encodes either the legacy
        // pixel-format index (0=BGRA, 1=GA8 — used by xcassets atlases)
        // OR the appearance-variant gamut (1 for the alternate atlas
        // emitted alongside the primary BGRA atlas on scrumdinger-shaped
        // .icon bundles). They never coexist in practice — the legacy
        // GA8 path always has gamut=0; the .icon variant atlases are
        // always BGRA — so we just take whichever is non-zero.
        let fmt_idx: u32 = if &self.pixel_format == b"BGRA" { 0 } else { 1 };
        let third = if self.gamut > 0 { self.gamut } else { fmt_idx };
        format!(
            "ZZZZPackedAsset-{}.{}.{}-gamut{}",
            self.scale, self.dim1, third, self.gamut
        )
    }

    pub fn bytes_per_row(&self) -> usize {
        let bpp: usize = if &self.pixel_format == b"BGRA" { 4 } else { 2 };
        let exact = self.width as usize * bpp;
        ((exact + 31) / 32) * 32
    }

    /// Blit all packed images into a single atlas pixel buffer. Rows are
    /// padded to 32-byte alignment to match the encoder's expected stride.
    pub fn render(&mut self) {
        let bpp: usize = if &self.pixel_format == b"BGRA" { 4 } else { 2 };
        let bpr = self.bytes_per_row();
        let mut buf = vec![0u8; bpr * self.height as usize];

        for img in &self.images {
            let src_stride = img.width as usize * bpp;
            for row in 0..img.height as usize {
                let src_off = row * src_stride;
                let dst_off = (img.y as usize + row) * bpr + img.x as usize * bpp;
                if src_off + src_stride <= img.pixel_data.len() {
                    buf[dst_off..dst_off + src_stride]
                        .copy_from_slice(&img.pixel_data[src_off..src_off + src_stride]);
                }
            }
        }
        self.pixel_data = buf;
    }
}

struct Column {
    x: u32,
    width: u32,
    bottom: u32,
}

struct Shelf {
    y: u32,
    height: u32,
    columns: Vec<Column>,
}

pub fn pack_images_split(
    mut images: Vec<PackedImage>,
    max_width: u32,
    max_height: u32,
) -> Vec<Atlas> {
    if images.is_empty() {
        return Vec::new();
    }
    images.sort_by(|a, b| b.height.cmp(&a.height).then(b.width.cmp(&a.width)));
    let mut remaining = images;
    let mut atlases = Vec::new();

    while !remaining.is_empty() {
        let first = &remaining[0];
        let mut atlas = Atlas {
            pixel_format: first.pixel_format,
            scale: first.scale,
            ..Default::default()
        };
        let overflow = pack_shelf_atlas(&mut atlas, remaining, max_width, max_height);
        atlases.push(atlas);
        remaining = overflow;
    }
    atlases
}

fn pack_shelf_atlas(
    atlas: &mut Atlas,
    sorted_imgs: Vec<PackedImage>,
    max_width: u32,
    max_height: u32,
) -> Vec<PackedImage> {
    let mut shelves: Vec<Shelf> = Vec::new();
    let mut atlas_width: u32 = 0;
    let mut placed: Vec<PackedImage> = Vec::new();
    let mut overflow: Vec<PackedImage> = Vec::new();

    for mut img in sorted_imgs {
        let mut fit = false;

        for (si, shelf) in shelves.iter_mut().enumerate() {
            let is_first_shelf = si == 0;

            // Try existing columns
            for col in shelf.columns.iter_mut() {
                if img.width <= col.width {
                    let new_bottom = col.bottom + GAP + img.height;
                    if new_bottom <= shelf.y + shelf.height {
                        img.x = col.x;
                        img.y = col.bottom + GAP;
                        col.bottom = img.y + img.height;
                        placed.push(img.clone());
                        fit = true;
                        break;
                    }
                }
            }
            if fit {
                break;
            }

            // New column on this shelf
            let new_x = if let Some(last) = shelf.columns.last() {
                last.x + last.width + GAP
            } else {
                MARGIN
            };

            if img.height <= shelf.height {
                let mut width_ok =
                    is_first_shelf && new_x + img.width + MARGIN <= max_width;
                if !width_ok && atlas_width > 0 {
                    width_ok = new_x + img.width + MARGIN <= atlas_width;
                }
                if !width_ok && shelf.columns.is_empty() {
                    width_ok = true;
                }

                if width_ok {
                    img.x = new_x;
                    img.y = shelf.y;
                    let bottom = img.y + img.height;
                    shelf.columns.push(Column {
                        x: new_x,
                        width: img.width,
                        bottom,
                    });
                    let new_right = new_x + img.width + MARGIN;
                    if new_right > atlas_width {
                        atlas_width = new_right;
                    }
                    placed.push(img.clone());
                    fit = true;
                    break;
                }
            }
        }
        if fit {
            continue;
        }

        // New shelf
        let new_y = if let Some(last) = shelves.last() {
            last.y + last.height + GAP
        } else {
            MARGIN
        };

        if new_y + img.height + MARGIN <= max_height || shelves.is_empty() {
            img.x = MARGIN;
            img.y = new_y;
            let shelf_height = img.height;
            let bottom = img.y + img.height;
            let new_right = MARGIN + img.width + MARGIN;
            if new_right > atlas_width {
                atlas_width = new_right;
            }
            let shelf = Shelf {
                y: new_y,
                height: shelf_height,
                columns: vec![Column {
                    x: MARGIN,
                    width: img.width,
                    bottom,
                }],
            };
            shelves.push(shelf);
            placed.push(img);
        } else {
            overflow.push(img);
        }
    }

    if !shelves.is_empty() {
        atlas.width = atlas_width;
        let mut max_bottom = 0;
        for shelf in &shelves {
            for col in &shelf.columns {
                if col.bottom > max_bottom {
                    max_bottom = col.bottom;
                }
            }
        }
        atlas.height = max_bottom + MARGIN;
    }
    atlas.images = placed;
    overflow
}

/// Group renditions into packable groups and inline-only renditions.
///
/// Returns `(pack_groups, inline_renditions)` where each pack group is
/// `(pixel_format, scale, rendition_indices)` — the caller keeps
/// ownership of the renditions vector, so we pass indices back.
pub fn group_for_packing(
    renditions: &[crate::car::Rendition],
) -> (Vec<([u8; 4], u16, Vec<usize>)>, Vec<usize>) {
    use crate::car;

    const ICON_INLINE_THRESHOLD: u32 = 256;
    const PACK_MAX_WIDTH: u32 = 262;
    const PACK_MAX_HEIGHT: u32 = 196;
    const PACK_MARGIN: u32 = 4;

    // group key: (pixel_format, scale, sprite_atlas_id) -> list of indices
    let mut groups: indexmap::IndexMap<([u8; 4], u16, u16), Vec<usize>> =
        indexmap::IndexMap::new();
    let mut force_inline: Vec<usize> = Vec::new();

    for (i, rend) in renditions.iter().enumerate() {
        match rend.layout {
            car::LAYOUT_MULTISIZE_IMAGE
            | car::LAYOUT_COLOR
            | car::LAYOUT_RAW_DATA
            | car::LAYOUT_METADATA => {
                force_inline.push(i);
                continue;
            }
            _ => {}
        }
        if rend.part == car::PART_ICON && rend.width >= ICON_INLINE_THRESHOLD {
            force_inline.push(i);
            continue;
        }
        if rend.csi_override.is_some() {
            force_inline.push(i);
            continue;
        }
        if rend.width >= PACK_MAX_WIDTH - PACK_MARGIN
            || rend.height >= PACK_MAX_HEIGHT - PACK_MARGIN
        {
            force_inline.push(i);
            continue;
        }
        let key = (rend.pixel_format, rend.scale, rend.sprite_atlas_id);
        groups.entry(key).or_default().push(i);
    }

    let mut pack_groups: Vec<([u8; 4], u16, Vec<usize>)> = Vec::new();
    let mut inline: Vec<usize> = force_inline;

    // Sort keys deterministically to match Python's sorted()
    let mut keys: Vec<_> = groups.keys().cloned().collect();
    keys.sort_by_key(|k| (k.0, k.1, k.2));
    for k in keys {
        let idxs = groups.shift_remove(&k).unwrap();
        let distinct: std::collections::HashSet<u16> =
            idxs.iter().map(|i| renditions[*i].identifier).collect();
        if distinct.len() >= 2 {
            pack_groups.push((k.0, k.1, idxs));
        } else {
            inline.extend(idxs);
        }
    }

    (pack_groups, inline)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn img(name: &str, w: u32, h: u32) -> PackedImage {
        PackedImage::new(name.to_string(), 0, w, h)
    }

    #[test]
    fn empty_packing() {
        let atlases = pack_images_split(vec![], 262, 196);
        assert!(atlases.is_empty());
    }

    #[test]
    fn single_image_placed_with_margin() {
        let atlases = pack_images_split(vec![img("a", 10, 10)], 262, 196);
        assert_eq!(atlases.len(), 1);
        let atlas = &atlases[0];
        assert_eq!(atlas.images.len(), 1);
        assert_eq!(atlas.images[0].x, MARGIN);
        assert_eq!(atlas.images[0].y, MARGIN);
        assert_eq!(atlas.width, MARGIN + 10 + MARGIN);
    }

    #[test]
    fn two_images_same_height_go_into_columns() {
        let atlases = pack_images_split(
            vec![img("a", 10, 20), img("b", 15, 20)],
            262,
            196,
        );
        assert_eq!(atlases.len(), 1);
        let atlas = &atlases[0];
        assert_eq!(atlas.images.len(), 2);
        // Both on the first shelf (y = MARGIN) side by side.
        assert!(atlas.images.iter().all(|i| i.y == MARGIN));
    }

    #[test]
    fn oversized_image_overflows_new_atlas() {
        // Ask for two atlases worth — first has a mega shelf, second shelf
        // won't fit when max_height is low.
        let atlases = pack_images_split(
            vec![
                img("tall1", 50, 100),
                img("tall2", 50, 100),
                img("tall3", 50, 100),
            ],
            262,
            196,
        );
        // Should split: a tall shelf of ~100 leaves ~90 for another shelf which fits one
        assert!(atlases.iter().map(|a| a.images.len()).sum::<usize>() == 3);
    }

    #[test]
    fn atlas_name_format() {
        let atlas = Atlas {
            scale: 2,
            dim1: 3,
            pixel_format: *b"BGRA",
            ..Default::default()
        };
        assert_eq!(atlas.name(), "ZZZZPackedAsset-2.3.0-gamut0");
        let atlas = Atlas {
            scale: 1,
            dim1: 0,
            pixel_format: *b" 8AG",
            ..Default::default()
        };
        assert_eq!(atlas.name(), "ZZZZPackedAsset-1.0.1-gamut0");
    }

    #[test]
    fn bytes_per_row_alignment() {
        let atlas = Atlas {
            width: 10,
            pixel_format: *b"BGRA",
            ..Default::default()
        };
        // 10 * 4 = 40, aligned to 32 is 64
        assert_eq!(atlas.bytes_per_row(), 64);

        let atlas = Atlas {
            width: 10,
            pixel_format: *b" 8AG",
            ..Default::default()
        };
        // 10 * 2 = 20, aligned to 32 is 32
        assert_eq!(atlas.bytes_per_row(), 32);
    }

    #[test]
    fn render_copies_pixels() {
        let mut atlas = Atlas {
            width: 10,
            height: 10,
            pixel_format: *b"BGRA",
            ..Default::default()
        };
        let mut pi = img("a", 2, 2);
        pi.x = 0;
        pi.y = 0;
        pi.pixel_data = vec![
            1, 2, 3, 4, 5, 6, 7, 8, // row 0
            9, 10, 11, 12, 13, 14, 15, 16, // row 1
        ];
        atlas.images = vec![pi];
        atlas.render();
        // First bytes of first row should be [1,2,3,4,5,6,7,8]
        assert_eq!(&atlas.pixel_data[..8], &[1, 2, 3, 4, 5, 6, 7, 8]);
        // Second row offset = bytes_per_row (64 for width=10 BGRA)
        let bpr = atlas.bytes_per_row();
        assert_eq!(&atlas.pixel_data[bpr..bpr + 8], &[9, 10, 11, 12, 13, 14, 15, 16]);
    }
}
